pkg/block/index.go

```diff
@@ -345,13 +347,17 @@ func GatherIndexIssueStats(logger log.Logger, fn string, minTime int64, maxTime
 		if lastLset != nil && labels.Compare(lastLset, lset) >= 0 {
 			return stats, errors.Errorf("series %v out of order; previous %v", lset, lastLset)
 		}
-		l0 := lset[0]
-		for _, l := range lset[1:] {
-			if l.Name <= l0.Name {
-				return stats, errors.Errorf("out-of-order label set %s for series %d", lset, id)
-			}
-			l0 = l
+
+		// Ensure that __name__ comes first.
+		if lset[0].Name != model.MetricNameLabel {
+			return stats, errors.Errorf("label set %s does not start with %s for series %d", lset, model.MetricNameLabel, id)
 		}
+
+		// Ensure all but the first value (__name__) are sorted.
+		if !sort.SliceIsSorted(lset[1:], func(i, j int) bool { return lset[i+1].Name <= lset[j+1].Name }) {
```

Comment:
It's a bit different and I might raise a different pull request but what if we would relax this check a bit and switch it to `<` here? Essentially, it seems like Prometheus accepts duplicate labels and if that happens (happened in our case) then compactor doesn't work anymore due to this check. Case in point:

```
{"caller":"main.go:181","err":"error executing compaction: compaction failed: compaction: gather index issues for block /data/compact/0@{monitor=\"monitor\",replica=\"repl\"}/01D34EDQMSQ29RHAC47XGKHGC7: out-of-order label set {__name__=\"foo\",exported_job=\"vv\",host=\"172_16_226_56\",host=\"172_16_226_56\",region=\"lt\",subtask_index=\"5\",task_attempt_id=\"32e4b047bb768583ff57c709be3b1046\",task_attempt_num=\"8\",task_id=\"688c028a219ff3372f3eecb0ee5811f9\",task_name=\"Source:_foo\",tenant=\"abc\",tier=\"cooltier\",tm_id=\"53b2ed987b08f427dec4ee1465df91fa\"} for series 2594231","level":"error","msg":"running command failed","ts":"2019-02-11T13:30:33.901722306Z"}
```

Reply:
Moved to https://github.com/improbable-eng/thanos/pull/848.